syndecrypt/__main__.py

```diff
@@ -60,22 +62,36 @@ def main(args):
     f = arguments['<encrypted-file>']
     ff = os.path.abspath(f)
     fp = os.path.basename(ff)
-    
+
     if os.path.isdir(ff):
         if not os.path.isdir(os.path.join(output_dir, fp)):
             output_dir = os.path.join(output_dir, fp)
             os.mkdir(output_dir)
         else:
             print("Folder already exists!")
-        for root, subdirs, items in os.walk(ff):
-            structure = root.replace(ff, output_dir, 1)
+
+        directories = list(os.walk(ff))
+
+        for input_dir, _, _ in directories:
+            structure = input_dir.replace(ff, output_dir, 1)
             if not os.path.isdir(structure):
                 os.mkdir(structure)
-            
-            for filename in items:
-                file_path = os.path.join(root, filename)
-                if filename != ".DS_Store":
-                    files.decrypt_file(file_path, os.path.join(structure, filename), password=password, private_key=private_key, public_key=public_key)
+
+        decrypt_args = []
+
+        for input_dir, _, filenames in directories:
+            for filename in filenames:
+                decrypt_args.append((
+                    os.path.join(input_dir, filename),
+                    os.path.join(input_dir.replace(ff, output_dir, 1), filename),
+                    password,
+                    private_key,
+                    public_key,
+                ))
+
+        with Pool() as p:
+            p.starmap(files.decrypt_file, decrypt_args)
+
```

Comment:
I've been testing these changes out with a `.sparsebundle` Time Machine backup which is a mix of files and folders and it works just the same as it did.

The code you originally had would make directories as it decrypted the files that needed. These changes split the loop so that it's two pass: first to create the directory hierarchy in serial then second to populate it with all the files in parallel. I did it this way so that the main decryption loop was simpler to parallelize given the signature of `decrypt_file`, which does not create any necessary directories.

So in short, the logic hasn't changed, it's just reordered.

As noted in the PR description, this implementation assumes that the source directory isn't being written to and that the destination directory is only being written to by this process. If anyone else interferes while it's running, all bets are off, but that was already true.

Reply:
Yes I see that you store the list of directories, make one pass to recreate the structure then another pass to create a list of files to be decrypted and execute the decryption call in parallel. Very nice work!